src/components/coin.ts

```diff
@@ -269,3 +274,191 @@ export const applyBonusByUserId = async (userId: string): Promise<boolean> => {
   }
   return false;
 };
+
+export enum TransferSign {
+  Pending = 0,
+  Accept = 1,
+  Decline = 2,
+}
+
+export const getEmojiFromSign = (sign: TransferSign): string => {
+  switch (sign) {
+    case TransferSign.Pending:
+      return '❓';
+    case TransferSign.Accept:
+      return '✅';
+    case TransferSign.Decline:
+      return '❌';
+  }
+};
+
+export enum TransferResult {
+  Pending,
+  Rejected,
+  Confirmed,
+}
+
+type TransferState = {
+  sender: User;
+  receiver: User;
+  result: TransferResult;
+  amount: number;
+  reason?: string;
+};
+
+class TransferTracker {
+  transfers: Map<string, Transfer>; // id, transfer
+
+  constructor() {
+    this.transfers = new Map<string, Transfer>();
+  }
+  getTransferFromId(id: string): Transfer | undefined {
+    return this.transfers.get(id);
+  }
+
+  runFuncOnGame(transferId: string, func: (transfer: Transfer) => void): void {
+    func(this.getTransferFromId(transferId)!);
+  }
+
+  async startTransfer(
+    sender: User,
+    receiver: User,
+    amount: number,
+    channelId: string,
+    client: SapphireClient<boolean>,
+    reason?: string,
+  ): Promise<Transfer> {
+    const transferId = uniqueId();
+    const transferState: TransferState = {
+      sender: sender,
+      receiver: receiver,
+      amount: amount,
+      reason: reason ?? '',
+      result: TransferResult.Pending,
+    };
+    const transfer = new Transfer(channelId, client, transferId, transferState);
+    this.transfers.set(transferId, transfer);
+    return transfer;
+  }
+
+  async endTransfer(transferId: string): Promise<void> {
+    const transfer = this.transfers.get(transferId);
+    if (!transfer) {
+      throw new Error(`No transfer with transfer ID ${transferId} found`);
+    }
+
+    if (transfer.state.result === TransferResult.Pending) return;
+    await transfer.handleTransaction();
+  }
+}
+
+export const transferTracker = new TransferTracker();
+
+export class Transfer {
+  channelId: string;
+  client: SapphireClient<boolean>;
+  state: TransferState;
+  transferId: string;
+  transferMessage!: SapphireSentMessageType;
+
+  constructor(
+    channelId: string,
+    client: SapphireClient<boolean>,
+    transferId: string,
+    transferState: TransferState,
+  ) {
+    this.channelId = channelId;
+    this.state = transferState;
+    this.client = client;
+    this.transferId = transferId;
+  }
+
+  // called if state is (believed to be) no longer pending. Transfers coins and updates balances if transfer is confirmed
+  async handleTransaction(): Promise<void> {
+    if (this.state.result === TransferResult.Confirmed) {
+      // Adjust the receiver balance with coins transferred
+      await adjustCoinBalanceByUserId(
+        this.state.receiver.id,
+        this.state.amount,
+        UserCoinEvent.AdminCoinAdjust,
```

Comment:
This should not be AdminCoinAdjust. Maybe make a new enum value `UserCoinEvent.CoinTransferAcceptor` for this?